particle-iot/device-os
# BLE: alternately broadcast user's and Particle-specific advertising d…

### Problem

The advertising data and scan response data set by user application will not be broadcasting when device in the Listening mode. What that means is, for custom mobile App it would not be able to filter their device by the advertising data set in user application when device in the Listening mode.

### Solution

Alternately broadcast user's and Particle-specific advertising data when device in the Listening mode. By this way, both of the advertising data will be broadcasting and observed by Particle mobile App and custom mobile App. The only defect is that when device in the listening mode, the advertising and connection parameters set in user application are adjusted to appropriate values to coordinate the Particle mobile setup. But once device exits the Listening mode, these parameters are restored for user application.

Effective advertising parameters in the Listening mode:
```
|<--------------->|<--------------->|<--------------->|<--------------->|
User ADV          Particle ADV      User ADV          Particle ADV 

+----------+---------+
| Interval | Timeout |
+----------+---------+
| 20ms     | 1000ms  |
+----------+---------+
```

Effective connection parameters in the Listening mode:
```
+--------------+--------------+---------+---------+
| MIN Interval | MAX Interval | Latency | Timeout |
+--------------+--------------+---------+---------+
| 30ms         | 50ms         | 0       | 5000ms  |
+--------------+--------------+---------+---------+
```

### Steps to Test

Build and flash the monolithic firmware with log enabled:
    `$ make MODULAR=n DEBUG_BUILD=y PLATFORM=xenon TEST=<path_to_the_example>`
    `$ dfu-util -d 2b04:d00e -a 0 -s 0x30000:leave -D <path_to_the_built_binary>`

1. Device is neither broadcasting nor connected before entering the Listening mode:
    - Check the log message when device in the Listening mode, it should be only broadcasting the Particle-specific advertising data.
    - It should be able to be connected by the Particle mobile App.
    - Make the device leave the listening mode and disconnect from mobile App, device shouldn't be broadcasting.

2. Device is broadcasting before entering the Listening mode:
    - Check the log message when device in the Listening mode, it should be alternately broadcasting user's and Particle-specific advertising data in every 500ms interval.
    - It should be able to be connected by the Particle mobile App.
    - Make the device leave the listening mode and disconnect from mobile App, device should be only broadcasting the user's advertising data with the advertising parameters set in user application.

3. Device is connected before entering the Listening mode:
    - Check the log message when device in the Listening mode, it shouldn't be broadcasting.
    - Disconnect device when device in the listening mode, it should be alternately broadcasting user's and Particle-specific advertising data in every 500ms interval.
    - It should be able to be connected by the Particle mobile App.
    - Make the device leave the listening mode and disconnect from mobile App, device should be only broadcasting the user's advertising data with the advertising parameters set in user application.

### Example App

```c
#include "Particle.h"

SYSTEM_MODE(MANUAL);

Serial1LogHandler log(115200, LOG_LEVEL_ALL);

const char* serviceUuid = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E";

void setup() {
    BLE.setAdvertisingTimeout(1000); // 1000 * 10 ms

    BleAdvertisingData data;
    data.appendServiceUUID(serviceUuid);
    BLE.advertise(&data);
}

void loop() {
}
```
